simpleviewinc/keg-cli
# ZEN-552: Labels

**Ticket**: [ZEN-552](https://jira.simpleviewtools.com/browse/ZEN-552)

> when this gets merged into develop, we need to push to master as well, so that the keg-deploy provision script can use it when cloning the master branch of keg-cli

## Context
* The Keg-CLI uses docker labels to allow automatic proxy urls through the keg-proxy
* Currently these labels expect the proxy domain to include the `local` subdomain
* When on the staging server, we'll need to update the labels to use `staging`

## Goal

* Update the labels to use the right environment whenever a container is run

## Updates
* `README.md`
  * added some additional documentation on configuration and architecture
* `**/*.config.json` && `**/*.env`
  * updated the `.env` setup files to set the `KEG_PROXY_HOST` env to `<defaultEnv>.<defaultDomain>`
  * updated the `*.config.json` files to define a `defaultDomain` field (set to `keghub.io`)
* `containers/proxy/values.yml`
  * updated default domain of this proxy override
* `package.json`
  * updated a jest alias for the keg envs. I needed these to be mocked in one of the unit tests I wrote.   
  * I wasn't able to mock it just by using `__mocks__` and `jest.mock` alone; it seems jest mocks don't work with aliases, unless you do it this way and just alias straight to the mock
* `src/__mocks__/constants/envs.js`
  * mocks for the env constants
* `src/tasks/docker/package/run.js`
  * this is the main update of this PR
  * updated the `run` task to override the traefik rule label if it is already set, using the current environment's env for the domain
* `src/utils/docker/getOptsWithLabels.js`
  * function that actually gets the right labels for the environment

## Testing

### Staging Server
* ssh onto the current ec2 instance
  * you will need to add your ip to the private sg
  * it will have this branch of the keg-cli already setup
* start a keg package
  * I recommend this retheme one: `keg d pack run package=ghcr.io/simpleviewinc/retheme:zen-526-upgrade-expo`
  * you can try others like `evf`, but it runs very **very** slow
    * we probably need to disable the nodemon listeners for packaged containers
    * we also probably should increase the ec2 performance tier at some point
* follow the URL in your browser
  * if you ran the aforementioned retheme package, go to: http://retheme-ZEN-526-upgrade-expo.staging.keghub.io
* verify you can reach that URL in your browser


### Local dev

#### Setup
* on your local machine, switch to this branch of your keg-cli
* modify your `.kegConfig/cli.config.json`
  * add `"defaultDomain": "keghub.io"` in the `settings` object
  * set your `settings.defaultEnv` value to `local`
* modify your `.kegConfig/defaults.env`
  * set `KEG_PROXY_HOST` like this:
```bash
KEG_PROXY_HOST={{ cli.settings.defaultEnv }}.{{ cli.settings.defaultDomain }}
```

#### Testing
* run any container (a local repo, not a package) that serves a web page
  * e.g. `keg evf start`
* verify that the application url matches the pattern: `<app_subdomain>.local.keghub.io`
* verify you can visit that url
